src/main/java/com/example/b2b_opportunities/Service/MailService.java

```diff
@@ -0,0 +1,51 @@
+package com.example.b2b_opportunities.Service;
+
+import com.example.b2b_opportunities.Entity.User;
+import com.example.b2b_opportunities.Exception.ServerErrorException;
+import jakarta.mail.MessagingException;
+import jakarta.mail.internet.MimeMessage;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.mail.javamail.MimeMessageHelper;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class MailService {
+
+    private final ConfirmationTokenService confirmationTokenService;
+    private final JavaMailSender mailSender;
+    @Value("${spring.mail.username}")
+    private String fromMail;
+
+    private String generateConfirmationLink(User user) {
+        String token = confirmationTokenService.generateConfirmationCode(user);
+        return "<a href=http://localhost:8082/api/auth/register/confirm?token=" + token + ">Confirm Email</a>";
```

Comment:
public String generateConfirmationLink(User user, HttpServletRequest request) {
    String token = confirmationTokenService.generateConfirmationCode(user);
    
    // Extract the dynamic base URL from the request
    String baseUrl = request.getScheme() + "://" + request.getServerName() + ":" + request.getServerPort() + request.getContextPath();;
    
    return "<a href=" + baseUrl + "/api/auth/register/confirm?token=" + token + ">Confirm Email</a>";
} 
You have to make sure you get the url dynamically.

Reply:
Thanks, the URL is now generated dynamically.